File: src/MediaOrganizer.Core/Configuration/PatternsStore.cs
namespace MediaOrganizer.Core.Configuration;

/// <summary>一条文件名日期解析模式（patterns.json）。结构见 SRS §5.3。</summary>
public sealed class PatternDefinition
{
    public string Name { get; set; } = "";
    public string Pattern { get; set; } = "";

    /// <summary>捕获组名 → 组号。支持键：year/month/day/hour/minute/second/timestamp。</summary>
    public Dictionary<string, int> GroupMapping { get; set; } = new();

    /// <summary>明确忽略的捕获组号（如前后缀 (.*) 组），解析与推断时跳过（FR-3.1）。</summary>
    public List<int> IgnoredGroups { get; set; } = [];

    /// <summary>时间戳位数：10（秒）/ 13（毫秒）/ 16（微秒×10，除以 1000）。</summary>
    public int? TimestampLength { get; set; }

    public bool Enabled { get; set; } = true;
    public double Weight { get; set; } = 1.0;
    public bool Builtin { get; set; }
}

/// <summary>patterns.json 顶层结构。</summary>
public sealed class PatternsFile
{
    public int Version { get; set; } = 1;
    public List<PatternDefinition> Patterns { get; set; } = [];
}

/// <summary>patterns.json 的读写。</summary>
public static class PatternsStore
{
    /// <summary>
    /// 文件不存在或解析失败 → 内置模式；文件存在且解析成功则尊重内容（含用户清空全部模式的空列表）。
    /// </summary>
    public static List<PatternDefinition> Load(string path, Action<string>? onError = null)
    {
        if (!File.Exists(path)) return GetBuiltinPatterns();
        var file = JsonFileStore.Load<PatternsFile>(path, onError);
        return file is null ? GetBuiltinPatterns() : file.Patterns;
    }

    public static void Save(string path, IReadOnlyList<PatternDefinition> patterns)
        => JsonFileStore.Save(path, new PatternsFile { Patterns = patterns.ToList() });

    /// <summary>内置模式：覆盖微信/小红书/OPPO/通用日期/紧凑日期/时间戳等常见场景。</summary>
    public static List<PatternDefinition> GetBuiltinPatterns() =>
    [
        new()
        {
            Name = "微信导出", Pattern = @"mm_export(\d{13})",
            GroupMapping = new() { ["timestamp"] = 1 }, TimestampLength = 13, Weight = 1.0, Builtin = true
        },
        new()
        {
            // 时分秒必须拆为 3 个 2 位捕获组：合成单个 6 位组会被当 hour 解析（143022）而抛异常
            Name = "小红书", Pattern = @"(\d{4})(\d{2})(\d{2})[-_](\d{2})(\d{2})(\d{2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3, ["hour"] = 4, ["minute"] = 5, ["second"] = 6 },
            Weight = 0.9, Builtin = true
        },
        new()
        {
            Name = "通用日期时间", Pattern = @"(\d{4})[-_.](\d{1,2})[-_.](\d{1,2})[T _](\d{1,2})[-_.:](\d{2})[-_.:](\d{2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3, ["hour"] = 4, ["minute"] = 5, ["second"] = 6 },
            Weight = 1.0, Builtin = true
        },
        new()
        {
            Name = "通用日期", Pattern = @"(\d{4})[-_.](\d{1,2})[-_.](\d{1,2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3 },
            Weight = 0.9, Builtin = true
        },
        new()
        {
            Name = "紧凑日期时间", Pattern = @"(\d{4})(\d{2})(\d{2})[ _]?(\d{2})(\d{2})(\d{2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3, ["hour"] = 4, ["minute"] = 5, ["second"] = 6 },
            Weight = 0.9, Builtin = true
        },
        new()
        {
            Name = "紧凑日期", Pattern = @"(\d{4})(\d{2})(\d{2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3 },
            Weight = 0.8, Builtin = true
        },
        new()
        {
            Name = "秒级时间戳", Pattern = @"(^|[^0-9])(\d{10})([^0-9]|$)",
            GroupMapping = new() { ["timestamp"] = 2 }, TimestampLength = 10,
            Weight = 0.7, Builtin = true
        },
        new()
        {
            // 同上：时分秒拆为 3 个 2 位捕获组
            Name = "OPPO 相机", Pattern = @"IMG_(\d{4})(\d{2})(\d{2})_(\d{2})(\d{2})(\d{2})",
            GroupMapping = new() { ["year"] = 1, ["month"] = 2, ["day"] = 3, ["hour"] = 4, ["minute"] = 5, ["second"] = 6 },
            Weight = 0.8, Builtin = true
        },
        new()
        {
            Name = "毫秒级时间戳", Pattern = @"(^|[^0-9])(\d{13})([^0-9]|$)",
            GroupMapping = new() { ["timestamp"] = 2 }, TimestampLength = 13,
            Weight = 0.7, Builtin = true
        },
        new()
        {
            Name = "微秒级时间戳", Pattern = @"(^|[^0-9])(\d{16})([^0-9]|$)",
            GroupMapping = new() { ["timestamp"] = 2 }, TimestampLength = 16,
            Weight = 0.6, Builtin = true
        }
    ];
}

File: src-flutter/lib/core/patterns_store.dart
/// 文件名日期解析模式（patterns.json），结构对应 Core/Configuration/PatternsStore.cs。
/// 键名 PascalCase 与桌面版一致：桌面版魔术工具编辑的 patterns.json 可直接复制到本应用。
library;

import 'app_config.dart' show JsonFileStore;

/// 一条文件名日期解析模式。
class PatternDefinition {
  final String name;
  final String pattern;

  /// 捕获组名 → 组号。支持键：year/month/day/hour/minute/second/timestamp。
  final Map<String, int> groupMapping;

  /// 明确忽略的捕获组号（如前后缀 (.*) 组），解析时跳过。
  final List<int> ignoredGroups;

  /// 时间戳位数：10（秒）/ 13（毫秒）/ 16（微秒×10）。
  final int? timestampLength;
  bool enabled;
  double weight;
  final bool builtin;

  PatternDefinition({
    required this.name,
    required this.pattern,
    Map<String, int>? groupMapping,
    List<int>? ignoredGroups,
    this.timestampLength,
    this.enabled = true,
    this.weight = 1.0,
    this.builtin = false,
  })  : groupMapping = groupMapping ?? {},
        ignoredGroups = ignoredGroups ?? [];

  Map<String, Object?> toJson() => {
        'Name': name,
        'Pattern': pattern,
        'GroupMapping': groupMapping,
        'IgnoredGroups': ignoredGroups,
        if (timestampLength != null) 'TimestampLength': timestampLength,
        'Enabled': enabled,
        'Weight': weight,
        'Builtin': builtin,
      };

  factory PatternDefinition.fromJson(Map<String, Object?> j) => PatternDefinition(
        name: (j['Name'] as String?) ?? '',
        pattern: (j['Pattern'] as String?) ?? '',
        groupMapping: ((j['GroupMapping'] as Map<Object?, Object?>?) ?? const {})
            .map((k, v) => MapEntry(k.toString(), (v as num).toInt())),
        ignoredGroups: ((j['IgnoredGroups'] as List<Object?>?) ?? const [])
            .whereType<num>()
            .map((n) => n.toInt())
            .toList(),
        timestampLength: (j['TimestampLength'] as num?)?.toInt(),
        enabled: (j['Enabled'] as bool?) ?? true,
        weight: (j['Weight'] as num?)?.toDouble() ?? 1.0,
        builtin: (j['Builtin'] as bool?) ?? false,
      );

  PatternDefinition copyWith({bool? enabled}) => PatternDefinition(
        name: name,
        pattern: pattern,
        groupMapping: Map.of(groupMapping),
        ignoredGroups: List.of(ignoredGroups),
        timestampLength: timestampLength,
        enabled: enabled ?? this.enabled,
        weight: weight,
        builtin: builtin,
      );
}

/// patterns.json 的读写与内置模式清单。
class PatternsStore {
  static const _versionKey = 'Version';
  static const _patternsKey = 'Patterns';

  /// 文件缺失或无有效内容时回退内置模式（与 C# PatternsStore.Load 一致）。
  static List<PatternDefinition> load(String path) {
    final file = JsonFileStore.load(path);
    if (file != null) {
      final list = file[_patternsKey];
      if (list is List<Object?> && list.isNotEmpty) {
        return list
            .whereType<Map<String, Object?>>()
            .map(PatternDefinition.fromJson)
            .toList();
      }
    }
    return getBuiltinPatterns();
  }

  static void save(String path, List<PatternDefinition> patterns) {
    JsonFileStore.save(path, {
      _versionKey: 1,
      _patternsKey: patterns.map((p) => p.toJson()).toList(),
    });
  }

  /// 内置模式：覆盖微信/小红书/OPPO/通用日期/紧凑日期/时间戳等常见场景。
  static List<PatternDefinition> getBuiltinPatterns() => [
        PatternDefinition(
          name: '微信导出',
          pattern: r'mm_export(\d{13})',
          groupMapping: {'timestamp': 1},
          timestampLength: 13,
          weight: 1.0,
          builtin: true,
        ),
        PatternDefinition(
          name: '小红书',
          pattern: r'(\d{4})(\d{2})(\d{2})[-_](\d{6})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3, 'hour': 4, 'minute': 5, 'second': 6},
          weight: 0.9,
          builtin: true,
        ),
        PatternDefinition(
          name: '通用日期时间',
          pattern: r'(\d{4})[-_.](\d{1,2})[-_.](\d{1,2})[T _](\d{1,2})[-_.:](\d{2})[-_.:](\d{2})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3, 'hour': 4, 'minute': 5, 'second': 6},
          weight: 1.0,
          builtin: true,
        ),
        PatternDefinition(
          name: '通用日期',
          pattern: r'(\d{4})[-_.](\d{1,2})[-_.](\d{1,2})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3},
          weight: 0.9,
          builtin: true,
        ),
        PatternDefinition(
          name: '紧凑日期时间',
          pattern: r'(\d{4})(\d{2})(\d{2})[ _]?(\d{2})(\d{2})(\d{2})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3, 'hour': 4, 'minute': 5, 'second': 6},
          weight: 0.9,
          builtin: true,
        ),
        PatternDefinition(
          name: '紧凑日期',
          pattern: r'(\d{4})(\d{2})(\d{2})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3},
          weight: 0.8,
          builtin: true,
        ),
        PatternDefinition(
          name: '秒级时间戳',
          pattern: r'(^|[^0-9])(\d{10})([^0-9]|$)',
          groupMapping: {'timestamp': 2},
          timestampLength: 10,
          weight: 0.7,
          builtin: true,
        ),
        PatternDefinition(
          name: 'OPPO 相机',
          pattern: r'IMG_(\d{4})(\d{2})(\d{2})_(\d{6})',
          groupMapping: {'year': 1, 'month': 2, 'day': 3, 'hour': 4, 'minute': 5, 'second': 6},
          weight: 0.8,
          builtin: true,
        ),
        PatternDefinition(
          name: '毫秒级时间戳',
          pattern: r'(^|[^0-9])(\d{13})([^0-9]|$)',
          groupMapping: {'timestamp': 2},
          timestampLength: 13,
          weight: 0.7,
          builtin: true,
        ),
        PatternDefinition(
          name: '微秒级时间戳',
          pattern: r'(^|[^0-9])(\d{16})([^0-9]|$)',
          groupMapping: {'timestamp': 2},
          timestampLength: 16,
          weight: 0.6,
          builtin: true,
        ),
      ];
}
